app/controllers/saml_completion_controller.rb

```diff
@@ -5,23 +5,35 @@
 class SamlCompletionController < ApplicationController
   # Pass the original service provider request to the main SamlIdpController#auth method
   # via a POST with form parameters replacing the url query parameters
+  before_action :verify_sp_session_exists
+
   def index
-    if sp_session.present?
-      request_url = URI(sp_session[:request_url])
-      path_year = request_url.path[-4..-1]
-      path_method = "api_saml_finalauthpost#{path_year}_url"
-      action_url = Rails.application.routes.url_helpers.send(path_method)
+    request_url = URI(sp_session[:request_url])
+    action_path = build_action_path(request_url.path)
 
+    if !action_path
+      render_not_found
```

Comment:
Nit: With an `if` / `else`, I think it's most readable if the initial `if` is the positive condition, i.e. changing this to `if action_path` and swapping the bodies.

Alternatively, if the idea with having this first is because it can be handled quickly and succinctly, could also consider an early return instead?

```rb
return render_not_found if !action_path
```

Reply:
good call, went with the early return